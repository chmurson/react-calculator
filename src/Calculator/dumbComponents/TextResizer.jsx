import React from 'react'
import PropTypes from 'prop-types'
import Measure from 'react-measure'

const FONT_SIZE_MOD = 2

export class TextResizer extends React.Component {
  static propTypes = {
    maxWidth: PropTypes.number.isRequired,
    maxFontSize: PropTypes.number.isRequired,
    minFontSize: PropTypes.number.isRequired,
    children: PropTypes.func.isRequired
  }

  constructor(props) {
    super(props)

    this.fontSize = this.props.maxFontSize
  }

  onInputAreaSizeChange = contentRect => {
    const inputAreaWidth = contentRect.bounds.width

    if (inputAreaWidth > this.props.maxWidth - this.props.maxFontSize && this.fontSize > this.props.minFontSize) {
      this.safelyApplyNewFontSize(-FONT_SIZE_MOD)
    }
    if (inputAreaWidth < this.props.maxWidth - this.props.maxFontSize * 3 && this.fontSize < this.props.maxFontSize) {
      this.safelyApplyNewFontSize(FONT_SIZE_MOD)
    }
  }

  setRef = ref => {
    this.ref = ref
    this.applyFontSize()
  }

  applyFontSize() {
    this.ref.style['font-size'] = `${this.fontSize}px`
  }

  safelyApplyNewFontSize(mod) {
    const newFontSize = this.fontSize + mod
    this.fontSize = Math.floor(Math.max(this.props.minFontSize, Math.min(this.props.maxFontSize, newFontSize)))
    this.applyFontSize()
  }

  render() {
    return (
      <Measure bounds onResize={this.onInputAreaSizeChange} innerRef={this.setRef}>
        {({ measureRef }) => this.props.children({ measureRef })}
      </Measure>
    )
  }
}
