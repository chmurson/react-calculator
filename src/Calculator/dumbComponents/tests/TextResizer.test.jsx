import React from 'react'
import { mount } from 'enzyme'
import { TextResizer } from '../TextResizer'

describe('CalculatorDumb', () => {
  let component
  beforeAll(() => {
    component = mount(
      <TextResizer maxFontSize={20} maxWidth={200} minFontSize={10}>
        {() => null}
      </TextResizer>
    )
  })

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot()
  })
})
