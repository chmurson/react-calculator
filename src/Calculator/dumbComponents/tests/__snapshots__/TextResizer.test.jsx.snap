// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculatorDumb matches snapshot 1`] = `
<TextResizer
  maxFontSize={20}
  maxWidth={200}
  minFontSize={10}
>
  <_class
    bounds={true}
    innerRef={[Function]}
    onResize={[Function]}
  >
    <Measure
      bounds={true}
      contentRect={
        Object {
          "bounds": Object {},
          "client": Object {},
          "entry": Object {},
          "margin": Object {},
          "offset": Object {},
          "scroll": Object {},
        }
      }
      measure={[Function]}
      measureRef={[Function]}
    />
  </_class>
</TextResizer>
`;
